subsys/nfc/tnep/ch/poller.c

```diff
@@ -0,0 +1,294 @@
+			      const struct nfc_tnep_poller_msg *msg,
+			      bool timeout)
+{
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(ch_buf,
+		      CONFIG_NFC_TNEP_CH_PARSER_BUFFER_SIZE);
+
+	if (!check_tnep_svc(param)) {
+		return;
+	}
+
+	LOG_INF("TNEP Connection Handover Service selected");
+
+	ch_tnep_ctrl.svc_selected = true;
+
+	if (timeout) {
+		err = nfc_tnep_poller_svc_deselect();
+		if (err) {
+			LOG_ERR("Deselecting Connetion Handover service failed %d",
+				err);
+			error_handler(err);
+		}
+	}
+
+	if ((msg->msg->record_count == 1) &&
+	    (msg->status == NFC_TNEP_STATUS_SUCCESS)) {
+		if (ch_tnep_ctrl.cb->request_msg_prepare) {
+			err = ch_tnep_ctrl.cb->request_msg_prepare();
+			if (err) {
+				LOG_ERR("Handover Request prepare error: %d",
+					err);
+				error_handler(err);
+			}
+		}
+
+		return;
+	}
+
+	if ((nfc_ndef_ch_rec_check(*msg->msg->record,
+				   NFC_NDEF_CH_REC_TYPE_HANDOVER_REQUEST))) {
+		err = request_msg_handle(msg->msg, &ch_buf);
+		if (err) {
+			error_handler(err);
+		}
+	}
+}
+
+static void tnep_svc_sent(const struct nfc_ndef_tnep_rec_svc_param *param,
+			  const struct nfc_tnep_poller_msg *rsp_msg,
+			  bool timeout)
+{
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(ch_buf,
+		      CONFIG_NFC_TNEP_CH_PARSER_BUFFER_SIZE);
+
+	if ((rsp_msg->msg->record_count == 1) &&
+	    (rsp_msg->status == NFC_TNEP_STATUS_SUCCESS)) {
+		LOG_DBG("Tag respond only with Status Record. Ending communication");
+		err = nfc_tnep_poller_svc_deselect();
+		if (err) {
+			error_handler(err);
+		}
+
+		return;
+	}
+
+	if (!check_tnep_svc(param) &&
+	   (!ch_tnep_ctrl.svc_selected)) {
+		return;
+	}
+
+	if (timeout) {
+		LOG_DBG("Service sent timeout");
+
+		err = nfc_tnep_poller_svc_deselect();
+		if (err) {
+			error_handler(err);
+		}
+
+		return;
+	}
+
+	LOG_INF("Received new Connection Handover service data");
+
+	if ((nfc_ndef_ch_rec_check(*rsp_msg->msg->record,
+				   NFC_NDEF_CH_REC_TYPE_HANDOVER_SELECT))) {
+		LOG_INF("Connection Handover Select message received");
+
+		err = select_msg_handle(rsp_msg->msg, &ch_buf);
+		if (!err) {
+			err = nfc_tnep_poller_svc_deselect();
+		}
+
+		if (err) {
+			error_handler(err);
+		}
+	}
+}
+
+static void tnep_svc_deselected(void)
+{
+	if (!ch_tnep_ctrl.svc_selected) {
+		return;
+	}
+
+	ch_tnep_ctrl.svc_selected = false;
+	ch_tnep_ctrl.request_received = false;
+	ch_tnep_ctrl.carrier_inactive = false;
+
+	LOG_INF("Connection Handover service deselected");
+}
+
+static struct nfc_tnep_poller_cb tnep_cb = {
+	.svc_selected = tnep_svc_selected,
+	.svc_sent = tnep_svc_sent,
+	.svc_deselected = tnep_svc_deselected,
+};
+
+int nfc_tnep_ch_service_init(struct nfc_tnep_ch_cb *cb)
+{
+	if (!cb) {
+		LOG_ERR("NULL callback structure");
+		return -EINVAL;
+	}
+
+	if ((!cb->request_msg_prepare) ||
+	    (!cb->request_msg_recv) ||
+	    (!cb->select_msg_recv)) {
+		LOG_ERR("Invalid callback set");
+		return -EINVAL;
+	}
+
+	ch_tnep_ctrl.cb = cb;
+
+	nfc_tnep_poller_cb_register(&tnep_cb);
```

Comment:
I assume that this callback is also registered in the application (main.c). Do you depend on the order of callback registration in order for this library to work?

Reply:
Yes it is in `nfc_poller`. The callback order doesn't matter here

Reply:
good